### PR TITLE
OO design for Phinx within an application

### PR DESCRIPTION
**Pull request by alexbowers-tecmark**
This is a nicer design (in my opinion) for web based migration

### CODE REVIEW

File: app/web.php [hunk 1]
```diff
@@ -39,45 +39,28 @@
...
-// Mapping of route names to commands.
-$routes = [
-    'status'   => 'getStatus',
-    'migrate'  => 'getMigrate',
-    'rollback' => 'getRollback',
-    ];
+            $output = call_user_func([$wrap, 'getMigrate'], $environment);
+            $error  = $wrap->getExitCode() > 0;
+            if($error) {
+                return $error;
```
Comment (shadowhand) [inline app/web.php]: this is effectively:

```
if ($error) {
    return true;
}
return true;
```

Comment (alexbowers-tecmark) [inline app/web.php]: doesnt $error contain a string if it is true?

**Comment by shadowhand** [comment]
What's the point of this? All your doing is replacing an example web app with a class that proxies some of `TextWrapper` methods.

File: app/web.php [hunk 2]
```diff
@@ -39,45 +39,28 @@
...
+            $wrap = new Phinx\Wrapper\TextWrapper($app);
 
-// Mapping of route names to commands.
-$routes = [
-    'status'   => 'getStatus',
-    'migrate'  => 'getMigrate',
-    'rollback' => 'getRollback',
-    ];
+            $output = call_user_func([$wrap, 'getMigrate'], $environment);
+            $error  = $wrap->getExitCode() > 0;
```
Comment (shadowhand) [inline app/web.php]: No, error is a boolean, as per this line.

**Comment by alexbowers-tecmark** [comment]
The point of it is that in your code you can simply call Phinx::migrate() now to perform a migration. It is just a wrapper around a wrapper, however it is clean to use. I will look into improvements if this sort of design is desired by the community.

File: app/web.php (hunk 2 above)
Comment (alexbowers-tecmark) [inline app/web.php]: Ah my mistake. I'll fix that then.

**Comment by shadowhand** [comment]
Calling static methods does not make a piece of code "OO". This offers no benefit over:

```php
$phinx = require 'vendor/app/phinx.php';
$wrap = new Phinx\Wrapper\TextWrapper($phinx);
$output = $wrap->getMigrate();
```

File: app/web.php [hunk 3]
```diff
@@ -39,45 +39,25 @@
...
-    $args = implode(', ', [var_export($env, true), var_export($target, true)]);
-    echo "DEBUG: $command($args)" . PHP_EOL . PHP_EOL;
-}
-echo $output;
+ class Phinx {
+        public static function migrate($environment = 'default') {
+            $app = require __DIR__ . '/phinx.php';
+            $wrap = new Phinx\Wrapper\TextWrapper($app);
+
+            $output = call_user_func([$wrap, 'getMigrate'], $environment);
```
Comment (shadowhand) [inline app/web.php]: Why is `call_user_func` necessary here?

File: app/web.php [hunk 4]
```diff
@@ -39,45 +39,25 @@
...
+            $error  = $wrap->getExitCode() > 0;
+            if($error) {
+                return $error;
```
Comment (shadowhand) [inline app/web.php]:
```php
if (true) {
  return true;
}
return true;
```

File: app/web.php [hunk 5]
```diff
@@ -39,45 +39,25 @@
...
-header('Content-Type: text/plain', true, $error ? 500 : 200);
-if ($debug) {
-    // Show what command was executed based on request parameters.
-    $args = implode(', ', [var_export($env, true), var_export($target, true)]);
-    echo "DEBUG: $command($args)" . PHP_EOL . PHP_EOL;
-}
-echo $output;
+ class Phinx {
+        public static function migrate($environment = 'default') {
+            $app = require __DIR__ . '/phinx.php';
```
Comment (shadowhand) [inline app/web.php]: Every method call will create a new instance of Phinx?

**Comment by alexbowers-tecmark** [comment]
This is just a quick POC to see whether people have any comments on this style. I plan to look into it more deeply tonight and actually make it cleaner and more usable.

File: app/web.php (hunk 3 above)
Comment (alexbowers-tecmark) [inline app/web.php]: It isn't, this is very rough code. I suppose a pull request was a bad idea, I should really have done a RFC instead. My mistake

File: app/web.php (hunk 5 above)
Comment (alexbowers-tecmark) [inline app/web.php]: In my final version of this which I hope to get some done tonight, I will be rewriting this to not require a new instance every time, and to instead use singletons

**Comment by shadowhand** [comment]
Feel free to create a `Phinx\Wrapper\Singleton` class or something, but please do not outright replace `app/web.php` with something completely unrelated.

I'm not opposed to your approach, but the current implementation is very sloppy and not placed correctly. With some clean up, it could be a very useful contribution!

**Comment by alexbowers-tecmark** [comment]
Okay. Will do. Apologies. I should have instead created an issue with a RFC. But i guess this had the same effect, again. Sorry. Will try to improve it tonight / some time this week

**Comment by MartinDesigner** [comment]
Regarding only to design, look at this stuff:
imghome
So, as a supplement, however as a main point try this amazing free app. Createer.com is an online software for easily designing online ads, responsive landing pages and print materials such as flyers, posters, covers, invitations and more. Createer lets you easily share your designs on social media, send them via email, download as HTML, image or PDF. You can even design UI mockups for mobile as for standard display (html, js, css always generating without any coding required from you). Createer also includes 3,000 predesigned templates (browse here: https://beta2.createer.com/Templates?ghi) and over 250,000 images (illustrations, photos and icons) to work with in its user-friendly drag’n’drop enviroment. And finally this amazing stuff is here: https://beta2.createer.com/?ghi

**Comment by robmorgan** [comment]
closing for now until I see more support from the community.

**Pull request closed without merging**